src/plugins/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,221 @@
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
+        .then(function (pattern) {
+          reply(pattern);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'POST',
+    handler: function (req, reply) {
+      const client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
+      const mappings = _.omit(_.mapValues(_.indexBy(req.payload.fields, 'name'), (value) => {
+        return value.mapping;
+      }), _.isUndefined);
+      indexPattern.fields = JSON.stringify(_.map(indexPattern.fields, (field) => {
+        return _.omit(field, 'mapping');
+      }));
+
+
+      client.create({
+        index: '.kibana',
+        type: 'index-pattern',
+        id: indexPattern.title,
+        body: indexPattern
+      }).then((patternResponse) => {
+        if (!isWildcard || _.isEmpty(mappings)) {
+          return patternResponse;
+        }
+        else {
+          return client.indices.exists({
+            index: indexPattern.title
+          }).then((matchingIndices) => {
+            if (matchingIndices) {
+              throw Boom.conflict('Cannot create an index template if existing indices already match index pattern');
+            }
+            else {
+              return client.indices.putTemplate({
+                order: 0,
+                create: true,
+                name: patternToTemplate(indexPattern.title),
+                body: {
+                  template: indexPattern.title,
+                  mappings: {
+                    _default_: {
+                      properties: mappings
+                    }
+                  }
+                }
+              });
+            }
+          }).catch((templateError) => {
+            return client.delete({
+              index: '.kibana',
+              type: 'index-pattern',
+              id: indexPattern.title
+            }).then(() => {
+              throw templateError;
+            }, () => {
+              throw new Error(`index-pattern ${indexPattern.title} created successfully but index template creation
+                failed. Failed to rollback index-pattern creation, must delete manually.`);
+            });
+          });
+        }
+      }).then(() => {
+        reply('success').statusCode = 201;
+      }).catch(function (error) {
+        reply(handleESError(error));
+      });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns/{id}',
+    method: 'PUT',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
```

Comment:
The core question here is whether POSTing a pattern should also create a template. I don't have a strong opinion there, but if we do decide to separate them I would prefer to expose a /api/template endpoint and allow the client to use that instead of tacking on a parameter to the index-pattern endpoint.

Let me try to explain the current code a bit more though. The client already has to opt-in to template creation. They do this by providing mappings for the fields in their pattern. You'll notice the code you linked to omits mappings, that's because they're stored in the template, not in .kibana. The documentation for this endpoint should say "Create an index pattern. Optionally provide mappings for your fields and we'll create an index template for you as well." A field in the payload might look like this: 

```
referrer: {
  count: 1,
  scripted: false,
  mappings: {
    type: string,
    doc_values: false,
    index: analyzed
  }
}
```

The schema for the index pattern returned from this endpoint is a bit different than the current index patterns. Client code that uses this API would need to support this new format. I did this intentionally so that the mappings from the template and the info from .kibana could be glued into a single cohesive REST resource. The definition of an index pattern in terms of this API is (old index pattern + mappings). But again, that's just my attempt to make things REST-y if we want to create a template when POSTing to /index-patterns.